Feat/2989 component summary

evwilkin opened the pull request:
Closes #2989 

This PR adds a component summary above the tabs at the top of each component page, which is pulled from the existing guidelines for those components.
- Adds a new file `components-data.json` which provides one location to pass through additional information for all components.
  - This file contains both a summary description for each component (removed from the design guidelines for those components) as well as links to component illustrations for each component (towards #2992)
- Updates component page in `mdx.js` to render the component summary above the tabs
- Adds a new `component-illustrations` directory with component illustration images (towards #2992)

Overflow of workflow:
- `v4`:
  - `components-data.json` contains supplemental component information across all components - currently summary and illustration fields
  - `patternfly-docs.config.js`: `components-data.json` gets imported as `componentsData` & passed into the documentation framework package.
- `documentation-framework`:
  - `webpack.base.config.js`: `componentsData` is defined & destructured, then attached to `process.env.componentsData` to be available throughout build process
  - `app.js`: `componentsData` pulled from `process.env.componentsData` and passed as prop into `MDXTemplate`
  - `mdx.js`: component's ID is used to pull the summary from `componentsData`, the string is converted to JSX and then rendered as a new component above the tabs on component pages.

patternfly-build commented:
PF4 preview: https://patternfly-org-pr-3138-v4.surge.sh/v4

nicolethoen commented on packages/v4/patternfly-docs/components-data.json:
```diff
@@ -0,0 +1,354 @@
+  },
+  "action-list": {
+    "illustration": "./images/component-illustrations/action-list.png",
+    "summary": "An <b>action list</b> is a group of actions with set spacing."
```
Is it more accurate to describe an action list as a group of actions, controls, or buttons with set spacing?

nicolethoen commented on packages/v4/patternfly-docs/components-data.json:
```diff
@@ -0,0 +1,354 @@
+  },
+  "alert-group": {
+    "illustration": "./images/component-illustrations/alert-group.png",
+    "summary": "An <b>alert group</b> stacks and positions 2 or more alerts, either in a layer over the main content of a page or inline with the page content. Alert groups should always rank alerts by age, stacking new alerts on top of old ones as they surface."
```
Functionally, another primary use case for Alert Groups is for creating a live region for alerts. Maybe the first sentence could say, `An alert group stacks and positions 2 or more alerts in a live region, either in a layer over the main content of a page or inline with the page content...`

nicolethoen commented on packages/v4/patternfly-docs/components-data.json:
```diff
@@ -0,0 +1,354 @@
+  },
+  "application-launcher": {
+    "illustration": "./images/component-illustrations/application-launcher.png",
+    "summary": "An <b>application launcher</b> is an optional utility menu item that allows a user to launch a separate web application in a new browser window."
```
Seems it's more of an option menu rather than an optional menu item

nicolethoen commented on packages/v4/patternfly-docs/components-data.json:
```diff
@@ -0,0 +1,354 @@
+  },
+  "input-group": {
+    "illustration": "./images/component-illustrations/input-group.png",
+    "summary": "An <b>input group</b> includes multiple input types, like a text input and a button, in one control."
```
I'd propose clarifying this further by saying something like:
`An input group groups multiple related controls or inputs together so they appear as one control.`

nicolethoen commented on packages/v4/patternfly-docs/components-data.json:
```diff
@@ -0,0 +1,354 @@
+  },
+  "menu-toggle": {
+    "illustration": "./images/component-illustrations/menu-toggle.png",
+    "summary": "The <b>menu toggle</b> component pairs with the <Link to='https://www.patternfly.org/v4/components/menu/design-guidelines'>menu</Link> component to create composable dropdown and select menus. Using a menu toggle with a menu enables you to create custom component configurations not supported by the standard dropdown or select components. However, from a design perspective, the dro ...
```
Could we note that the menu toggle component can `pair with the Menu component OR the Panel component to create more customizable dropdown and select implementations`?

I also wonder if the sentence about the design perspective is appropriate in this section?

nicolethoen commented on packages/v4/patternfly-docs/components-data.json:
```diff
@@ -0,0 +1,354 @@
+  },
+  "toolbar": {
+    "illustration": "./images/component-illustrations/toolbar.png",
+    "summary": "A <b>toolbar</b> allows a user to manage and manipulate a data set. Data can be presented in any valid presentation, a table, a list, or a data visualization (chart), for example. The PatternFly Toolbar component is a flexible layout system that accommodates a variety of configurations that can be adapted to your specific needs."
```
Rather than the second sentence here, I think it should say that `the toolbar responsively accommodates controls and displays applied filters in chip groups.` or something like that

nicolethoen reviewed (commented):
not sure if we want to update copy before we merge this. It appears to be working great! So if we want to address copy in a follow up, then that's fine.

evwilkin commented:
@mmenestr @mcarrano the following components don't have any content left in the design guidelines with the summary removed, in these cases do we want to remove the design guidelines tab completely?
- menu-toggle
- options menu
- simple list
- tab content

@mmenestr additionally, the `Progress stepper` doesn't have any design guidelines currently (do we want to add a quick summary as part of this PR?) and what do you think about Nicole's suggestions above?

mcarrano commented:
@evwilkin this is looking great. In reference to your questions, currently when a component is missing design guidelines the tab does appear with this placeholder text, "We’re working on some guidelines for this component. If you have any questions in the meantime, feel free to [get in touch with us.](https://www.patternfly.org/v4/community#get-in-touch)." That is something Zack set up a while ago. Any reason we can't still have that text display until we can add a design guidelines page?

As for the two components that do not have design guidelines yet, they are both relatively new and this raises the question of how and when these get created as new components get added. I suspect they might be missing illustrations also? It would be simple to come up for some text for these now, but I would like to consider the larger process question. Maybe to put the question another way, who do you expect will be responsible for maintaining the JSON file with this information, @evwilkin ? Happy to discuss.

mmenestr commented:
@evwilkin the progress stepper guidelines are currently being worked on and there will be a PR up for those very soon. For the rest of them, we can just do what @mcarrano said and used the automated sentence Zack had set up.

As per Nicole's suggestion, I think they're all valid, and I think why not edit them while we're here yeah!

mmenestr commented:
One general comment I have is - can we make it so that there's less space between the header and the description text? I see that it's implemented as two separate sections right now, but could we maybe drop the top padding on the description section so it's just 24px and not 48px total?

evwilkin commented:
Updated per PR feedback:
- Updated several component summaries based on suggestions
- Do not show fallback default design guidelines text if no guidelines exist, instead just do not show that tab
- Remove top padding from component summary text to reduce space from component title

mcarrano reviewed (changes requested):
This is looking really good @evwilkin . Just a couple of things that I am noticing.

* Sidebar is missing a description.
* Tab content has a design guidelines tab, but no content on that tab.

Also looks like we are still missing descriptions for Progress stepper and Timestamp.

evwilkin commented:
@mcarrano made changes based on your feedback, open to any edits on the new component descriptions 😬

mmenestr commented on packages/v4/patternfly-docs/content/design-guidelines/components/dropdown/dropdown.md:
```diff
@@ -3,8 +3,6 @@ id: Dropdown
 section: components
 ---
 
-A **dropdown** presents a menu of actions or links in a constrained space that will trigger a process or navigate to a new location. See [select component guidelines](/components/select/) for more information about making one or more selections from a list of items in a value list.
```
For this one, can we link to the design guidelines tab, instead of just the select page, because we're specifically saying "guidelines"

mmenestr commented on packages/v4/patternfly-docs/content/design-guidelines/components/label/label.md:
```diff
@@ -3,8 +3,6 @@ id: Label
 section: components
 ---
 
-The **label** component allows users to add specific element captions for user clairty and convenience
```
Just noticed there's a typo in this one - should say "clarity" instead of "clairty"

mmenestr commented on packages/v4/patternfly-docs/content/design-guidelines/components/skeleton/skeleton.md:
```diff
@@ -3,8 +3,6 @@ id: Skeleton
 section: components
 ---
 
-A **skeleton** is a type of loading state that allows you to expose content incrementally. For content that may take a long time to load, use a [progress bar](/components/progress/react) in place of a skeleton.
```
progress bar link is dead because of the extra "react" at the end of the link

evwilkin replied:
Thanks for calling out the fix needed!

mmenestr commented on packages/v4/patternfly-docs/content/design-guidelines/components/text-input-group/text-input-group.md:
```diff
@@ -3,8 +3,6 @@ id: Text input group
 section: components
 ---
 
-A **text input group** is a more flexible composable version of a [text input](/components/text-input). It enables consumers of PatternFly to build custom inputs for filtering and similar use cases by placing elements like icons, [chip groups](/components/chip-group) and [buttons](/components/button) within a text input.
```
For some reason the chip groups and button links are dead for me. Looks like the link is directing to : /components/text-input-group/components/button, instead of just /components/button

evwilkin replied:
Thanks - they were missing the leading slash before `components` in the link

mmenestr reviewed (commented):
The tab content and the timestamp pages have no tabs at all - is that expected?

Left a few other comments... otherwise looks great !!

evwilkin commented:
@mmenestr thanks for combing through those descriptions, they weren't changed from what we had previously but it's great that we were able to catch those existing errors as part of this work.

- For the tab content & timestamp components not showing any tabs, it looks like that was intentionally written into the code to hide the tab if there was only 1 tab.  That said, now that it's potentially more common (since there's no default Design guidelines tab added) I agree that it looks incorrect, and aside from being inconsistent can make it a bit confusing to know which code you're looking at.  I've updated it so it will always show tabs on component pages, even if there's only 1 to show.

mmenestr commented:
Just looked again and now it looks like none of the pages have any tabs? All the other fixes look good though!

mcarrano reviewed (approved):
Thanks @evwilkin . Other than the issues that @mmenestr flagged, I think this looks great.
